runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java

```diff
@@ -155,10 +156,18 @@ public void setTimer(TimerData timerData) {
   @Override
   public void deleteTimer(
       StateNamespace namespace, String timerId, String timerFamilyId, TimeDomain timeDomain) {
-    throw new UnsupportedOperationException("Canceling a timer by ID is not yet supported.");
+    TimerData removedTimer = existingTimers.remove(namespace, timerId + '+' + timerFamilyId);
+    if (removedTimer != null) {
+      Preconditions.checkState(
+          removedTimer.getDomain().equals(timeDomain),
+          "%s doesn't match time domain %s of timer",
+          timeDomain,
+          removedTimer.getDomain());
```

Comment:
I don't really have context here, but I wonder if we should do this check first, and unconditionally, so that we don't end up with a partially removed timer?

Reply:
I did that initially, but removed the additional lookup again again as a timer is uniquely identified by `namespace` + `timerId`. I added this as sanity check here to detect any broken state, but as far as I understand this should never be the case.